keras_nlp/models/gpt_neo_x/gpt_neo_x_attention.py

```diff
@@ -0,0 +1,176 @@
+
+        # Output.
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            bias_axes="d",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="attention_output",
+        )
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+        if attention_mask is not None:
+            mask_expansion_axis = -3
+            for _ in range(
+                attention_scores.shape.rank - attention_mask.shape.rank
+            ):
+                attention_mask = tf.expand_dims(
+                    attention_mask, axis=mask_expansion_axis
+                )
+        return self._softmax(attention_scores, attention_mask)
+
+    def _compute_attention(
+        self, query, key, value, attention_mask=None, training=None
+    ):
+        attention_scores = tf.einsum("aecd,abcd->acbe", key, query)
+        attention_scores /= tf.sqrt(self.attn_head_size)
+
+        attention_scores = self._masked_softmax(
+            attention_scores, attention_mask
+        )
+        attention_scores = self._attn_dropout_layer(
+            attention_scores, training=training
+        )
+        attention_output = tf.einsum("acbe,aecd->abcd", attention_scores, value)
+
+        return attention_output, attention_scores
+
+    def call(
+        self,
+        hidden_states,
+        attention_mask,
+        return_attention_scores=False,
+        training=None,
+    ):
+        query_key_value = self._qkv_dense(hidden_states)
+
+        query = query_key_value[..., : self.attn_head_size]
+        key = query_key_value[
+            ..., self.attn_head_size : 2 * self.attn_head_size
+        ]
+        value = query_key_value[..., 2 * self.attn_head_size :]
+
+        query, key = self.rotary_embedding(query, key)
+
+        attention_output, attention_scores = self._compute_attention(
+            query=query,
+            key=key,
+            value=value,
+            attention_mask=attention_mask,
+            training=training,
+        )
+
+        # Reshape `attention_output` to `(batch_size, sequence_length, hidden_dim)`.
+        attention_output = tf.reshape(
+            attention_output,
+            [
+                tf.shape(attention_output)[0],
+                tf.shape(attention_output)[1],
+                self.hidden_dim,
+            ],
+        )
+
+        attention_output = self._output_dense(attention_output)
+
+        if return_attention_scores:
```

Comment:
we can probably just ditch this argument, no need for it, and this is not a general layer